executor/insert_test.go

```diff
@@ -1510,68 +1510,89 @@ func TestIssue32213(t *testing.T) {
-			"create table t (c int unique key)",
-			"replace into t values (1)",
-		},
-		{
-			"insert-ingore-pk",
-			"create table t (c int primary key clustered)",
-			"insert ignore into t values (1)",
-		},
-		{
-			"insert-ingore-uk",
-			"create table t (c int unique key)",
-			"insert ignore into t values (1)",
-		},
-		{
-			"insert-update-pk",
-			"create table t (c int primary key clustered)",
-			"insert into t values (1) on duplicate key update c = values(c)",
-		},
-		{
-			"insert-update-uk",
-			"create table t (c int unique key)",
-			"insert into t values (1) on duplicate key update c = values(c)",
-		},
-	} {
-		t.Run(tt.name, func(t *testing.T) {
-			tk1.MustExec("drop table if exists t")
-			tk1.MustExec(tt.ddl)
-			tk1.MustExec("insert into t values (1)")
-			tk1.MustExec("begin")
-			tk1.MustExec(tt.dml)
-			done := make(chan struct{})
-			go func() {
-				tk2.MustExec("delete from t")
-				done <- struct{}{}
-			}()
-			select {
-			case <-done:
-				require.Failf(t, "txn2 is not blocked by %q", tt.dml)
-			case <-time.After(100 * time.Millisecond):
-			}
-			tk1.MustExec("commit")
-			<-done
-			tk1.MustQuery("select * from t").Check([][]interface{}{})
-		})
+	for _, shouldLock := range []bool{false} {
+		for _, tt := range []struct {
+			name          string
+			ddl           string
+			dml           string
+			isClusteredPK bool
+		}{
+			{
+				"replace-pk",
+				"create table t (c int primary key clustered)",
+				"replace into t values (1)",
+				true,
+			},
+			{
+				"replace-uk",
+				"create table t (c int unique key)",
+				"replace into t values (1)",
+				false,
+			},
+			{
+				"insert-ignore-pk",
+				"create table t (c int primary key clustered)",
+				"insert ignore into t values (1)",
+				true,
+			},
+			{
+				"insert-ignore-uk",
+				"create table t (c int unique key)",
+				"insert ignore into t values (1)",
+				false,
+			},
+			{
+				"insert-update-pk",
+				"create table t (c int primary key clustered)",
+				"insert into t values (1) on duplicate key update c = values(c)",
+				true,
+			},
+			{
+				"insert-update-uk",
+				"create table t (c int unique key)",
+				"insert into t values (1) on duplicate key update c = values(c)",
+				false,
+			},
+		} {
+			t.Run(
+				tt.name+"-"+strconv.FormatBool(shouldLock), func(t *testing.T) {
+					tk1.MustExec(fmt.Sprintf("set @@tidb_lock_unchanged_keys = %v", shouldLock))
+					tk1.MustExec("drop table if exists t")
+					tk1.MustExec(tt.ddl)
+					tk1.MustExec("insert into t values (1)")
+					tk1.MustExec("begin")
+					println("t1 begin")
+					tk1.MustExec(tt.dml)
+					println("after t1 dml")
+					errCh := make(chan error)
+					go func() {
+						_, err := tk2.Exec("insert into t values (1)")
+						errCh <- err
+					}()
+					select {
+					case <-errCh:
+						println("done")
+						if shouldLock {
+							require.Failf(t, "txn2 is not blocked by %q", tt.dml)
+						}
+						close(errCh)
+					case <-time.After(100 * time.Millisecond):
```

Comment:
Perhaps using a delay of 200ms or 300ms would provide greater stability, but it would also take more time. It would be ideal if there was a way to verify the actual number of keystrokes being written, but I guess that such functionality does not currently exist.